app/code/Magento/Catalog/Model/Product/Url.php

```diff
@@ -162,9 +162,9 @@ public function getUrl(\Magento\Catalog\Model\Product $product, $params = [])
                     \Magento\Store\Model\ScopeInterface::SCOPE_STORE
                 );
 
-                if ($categoryId) {
+                if ($useCategories && $categoryId) {
                     $filterData[UrlRewrite::METADATA]['category_id'] = $categoryId;
-                } elseif (!$useCategories) {
+                } else {
                     $filterData[UrlRewrite::METADATA]['category_id'] = '';
```

Comment:
Please rewrite as ternary operator instead.

**QA Note**: https://github.com/magento/magento2/commit/f47347fdc2e92668ad5ea40a5e6ec9fee2238b0e issue needs to be rechecked

Reply:
@andrewbess also, please implement an integration test for corresponding scenario.

Reply:
@orlangur Ok. Thank you.